doc/ni_by_example_2.md

```diff
@@ -0,0 +1,395 @@
+
+
+##Horizontal Scaling
+Note that you will need sufficient processing cores to effectively horizontally scale. If your computer has 2 cores and you call `S8`, it may slow your work down, as `ni` tries to spin up more processes than your machine can bear.
+
+* `S`: Horizontal Scaling 
+  * `$ ni <data> S<# cores>[...]`: distributes the computation of `...` across `<# cores>` processors.
+
+##Advanced Perl Operations
+* `p'^{BEGIN_BLOCK} ...'`: Begin block
+  * A begin block is indicated by attaching a caret (`^`) to a block of code (encolsed in `{ }`). Outside of begin blocks, the Perl code is evaluated for every row; inside a begin block, the code is evaluated once and this evaluation takes precedence over 
+  * Begin blocks are useful for converting data closures to Perl data structures, and defining other constants that are used in 
+  These blocks are evaluated in their entirety 
+* `p'%h = <col_1><col_2>_ @lines`: Hash constructor
+  * Hash constructors are useful for filtering large datasets without having to invoke an expensive sort or an HDFS join. Hash constructors are useful inside of begin blocks, often using the following workflow:
+    * Generate a list of things you want to filter, and put it in a data closure. `::ids[list_of_ids]`
+    * Convert the data closure to a hash using a begin block (`^{%id_hash = ab_ ids}`)
+    * Filter another dataset (`ids_and_data`) using the hash (`exists($id_hash{a()})`)
+    * `ni ::ids[list_of_ids] ids_and_data rp'^{%id_hash = ab_ ids} exists($id_hash{a()})'` 
+
+
+##Advanced Perl Reducers
+###Buffered Readahead
+These operations are good for reducing 
+
+* `rw`: read while
+  * `@lines = rw {condition}`: read lines while a condition is met
+* `ru`: read until
+  * `@lines = ru {condition}`: read lines until a condition is met
+* `re`: read equal
+  * `@lines = re {condition}`: read lines while the value of the condition is equal.
+
+###Line-Array Reducers
+These operations can be used to reduce the data output by the readahead functions. Look at the input provided by the first perl statement, 
+
+* `ni n1p'cart ["a", "b", "c"], [1, 2]' p'sum b_ re {a}'`
+* `ni n1p'cart ["a", "b", "c"], [1, 2]' p'sum a_ re {b}'`
+* `ni n1p'cart ["a", "b", "c"], [1, 2]' p'r all {a_($_)} re {b}'`
+* `ni n1p'cart ["a", "a", "b", "c"], [1, 2]' p'r uniq a_ re {b}'`
+* `ni n1p'cart ["a", "b", "c"], [1, 2]' p'r maxstr a_ re {b}'`
+* `ni n1p'cart ["a", "b", "c"], [1, 2]' p'r reduce {$_ + a} 0, re{b}` <- DOES NOT WORK BECAUSE BILOW WROTE IT WRONG
+
+##Stream Splitting/Joining/Duplication
+I don't find these operators particularly useful in practice (with the exception of `=\>` for writing a file in the middle of a long processing pipeline), but it's possible that you will! Then come edit these docs and explain why.
+
+* `+`: append a stream to this one
+* `^`: prepend a stream to this one
+* `%`: duplicate this stream through a process, and include output
```

Comment:
I actually cannibalized this to interleave streams. The original `%` no longer exists, in part because it was so useless.